### PR TITLE
[WIP] Issue 2895:  Enable TableStore iterator apis on SegmentHelper.

### LINKED ISSUE #2895: Client and Wire Protocol for Tables
**Problem description**
Implement Client code and supporting Wire Protocol for Tables. Contracts defined in #2876 .
- TableSegment
- ClientFactory

**Problem location**
Client and Wire Protocol.

**Suggestions for an improvement**

### PR DESCRIPTION
**Pull request by shrids**
Note: This PR depends on PR #3312 for `io.pravega.common.util.ContinuationTokenAsyncIterator`

**Change log description**  
Enable TableStore iterator apis on the SegmentHelper.

**Purpose of the change**  
Fixes #2895 

**What the code does**  
This PR adds helper methods to enable the controller to access the table store iterator apis.

**How to verify it**  
All the newly added tests and existing tests should continue to pass.

### CODE REVIEW

**Comment by shrids** [comment]
[Issue 2895: Enable TableStore iterator apis on SegmentHelper](https://trello.com/c/95H650Mv/127-issue-2895-enable-tablestore-iterator-apis-on-segmenthelper)

File: client/src/main/java/io/pravega/client/stream/impl/Controller.java [hunk 1]
```diff
@@ -42,6 +44,14 @@
      */
     CompletableFuture<Boolean> createScope(final String scopeName);
 
+    /**
+     * Gets an async iterator on streams in scope.
+     *
+     * @param scopeName The name of the scope for which to list streams in.
+     * @return A future which when completed will contain an iterator for all Streams in the scope.
```
Comment (andreipaduroiu) [inline client/src/main/java/io/pravega/client/stream/impl/Controller.java]: This doesn't return a future

File: common/src/main/java/io/pravega/common/util/BlockingAsyncIterator.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+ * This class is a wrapper over async iterator that implements java.util.Iterator interface.
+ * This is blocking wrapper on async iterator's getNext method which implements blocking hasNext and next methods of 
+ * iterator interface. 
+ * This class is threadsafe but that has the drawback of blocking on asyncIterator's getNext within a lock. 
+ */
+@ThreadSafe
+@VisibleForTesting
+public class BlockingAsyncIterator<T> implements Iterator<T> {
+    private final AsyncIterator<T> asyncIterator;
+    private T next;
```
Comment (andreipaduroiu) [inline common/src/main/java/io/pravega/common/util/BlockingAsyncIterator.java]: GuardedBy("$lock")

File: common/src/main/java/io/pravega/common/util/BlockingAsyncIterator.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    private boolean canHaveNext;
```
Comment (andreipaduroiu) [inline common/src/main/java/io/pravega/common/util/BlockingAsyncIterator.java]: GuardedBy("$lock")

File: common/src/main/java/io/pravega/common/util/BlockingAsyncIterator.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    public BlockingAsyncIterator(AsyncIterator<T> asyncIterator) {
+        this.canHaveNext = true;
+        this.next = null;
+        this.asyncIterator = asyncIterator;
+    }
+
+    @Synchronized
+    private void load() {
+        if (next == null && canHaveNext) {
+            next = asyncIterator.getNext().join();
```
Comment (andreipaduroiu) [inline common/src/main/java/io/pravega/common/util/BlockingAsyncIterator.java]: Waiting on a future while holding a lock is never a good idea. Here's my suggestion for avoiding this:
- make `next` a `CompletableFuture<T>` and still guard it by a lock
- In this method
    - While holding the lock, check `next` for null, and if so, then set it to `asyncInterator.getNext()` (no `join`)
    - After releasing the lock, invoke `join` on `next` and inspect its result.

You may have to do a bit more bookkeeping (in this method) to track whether you actually invoked `getNext` or not.

I think you can also get rid of `canHaveNext`. If `next` is completed (it should always be once this method exits), if its contents is `null`, then `hasNext()` should return `false` and `next()` should throw.

File: common/src/main/java/io/pravega/common/util/ContinuationTokenAsyncIterator.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+@ThreadSafe
+@Slf4j
+/**
+ * This is a continuation token based async iterator implementation. This class takes a function that when completed will 
+ * have next batch of results with continuation token. 
+ * This class determines when to call the next iteration of function (if all existing results have been exhausted) and 
+ * ensures there is only one outstanding call. 
+ */
+public class ContinuationTokenAsyncIterator<T, U> implements AsyncIterator<U> {
```
Comment (andreipaduroiu) [inline common/src/main/java/io/pravega/common/util/ContinuationTokenAsyncIterator.java]: Please use more specific names for the type args. `T` and `U` are not very indicative.

Also, doesn't checkstyle complain about not having `@param` args for these type params?

File: common/src/main/java/io/pravega/common/util/ContinuationTokenAsyncIterator.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    private final Object lock = new Object();
+    @GuardedBy("lock")
+    @VisibleForTesting
+    @Getter(AccessLevel.PACKAGE)
```
Comment (andreipaduroiu) [inline common/src/main/java/io/pravega/common/util/ContinuationTokenAsyncIterator.java]: The Lombok getter is not thread safe. Write your own getter where you take care of this.

Also you should make a copy of it in your getter. Even if you return the reference to the queue, the contents can still be iterated and/or modified externally, without holding the lock.

File: common/src/main/java/io/pravega/common/util/ContinuationTokenAsyncIterator.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    private final Queue<U> queue;
+    @GuardedBy("lock")
+    @VisibleForTesting
+    @Getter(AccessLevel.PACKAGE)
+    private final AtomicReference<T> token;
```
Comment (andreipaduroiu) [inline common/src/main/java/io/pravega/common/util/ContinuationTokenAsyncIterator.java]: Why do you have an `AtomicReference` that is guarded by a synchronized lock?

And why do you expose the whole `AtomicReference` to the outside via a getter? You should have explicit methods that access only its contents.

File: common/src/main/java/io/pravega/common/util/ContinuationTokenAsyncIterator.java [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    private final Function<T, CompletableFuture<Map.Entry<T, Collection<U>>>> function;
+    @GuardedBy("lock")
+    private CompletableFuture<Void> outstanding;
+    private final AtomicBoolean canContinue;
+    private final AtomicBoolean isOutstanding;
+    
+    public ContinuationTokenAsyncIterator(Function<T, CompletableFuture<Map.Entry<T, Collection<U>>>> function, T tokenIdentity) {
```
Comment (andreipaduroiu) [inline common/src/main/java/io/pravega/common/util/ContinuationTokenAsyncIterator.java]: 1. Check for nulls in the args.
2. Javadoc for constructor

File: common/src/main/java/io/pravega/common/util/ContinuationTokenAsyncIterator.java [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        this.isOutstanding = new AtomicBoolean(false);
+    }
+
+    @Override
+    public CompletableFuture<U> getNext() {
+        final T continuationToken;
+        boolean toCall = false;
+        synchronized (lock) {
+            // if the result is available, return it without making function call
+            if (!queue.isEmpty()) {
```
Comment (andreipaduroiu) [inline common/src/main/java/io/pravega/common/util/ContinuationTokenAsyncIterator.java]: So you check this queue here and if it's not empty, you return the first element, else you invoke the function and add to the queue and reinvoke this method.

I think you can greatly simplify this class by removing the queue altogether and invoking your function (in a loop, maybe) and returning the result. You can also get rid of a few other fields as well.

File: common/src/test/java/io/pravega/common/util/IteratorTest.java [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+import java.util.stream.IntStream;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.atLeast;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verify;
+
+public class IteratorTest {
```
Comment (andreipaduroiu) [inline common/src/test/java/io/pravega/common/util/IteratorTest.java]: This should be named `ContinuationTokenIteratorTest`

File: controller/src/main/java/io/pravega/controller/server/SegmentHelper.java [hunk 11]
```diff
@@ -1035,6 +1044,199 @@ public void authTokenCheckFailed(WireCommands.AuthTokenCheckFailed authTokenChec
...
+                                                                                    final ConnectionFactory clientCF,
+                                                                                    final String delegationToken,
+                                                                                    final long clientRequestId) {
+
+        final Controller.NodeUri uri = getSegmentUri(scope, stream, 0L, hostControllerStore);
+        final String qualifiedName = getScopedStreamName(scope, stream);
+        final WireCommandType type = WireCommandType.READ_TABLE_KEYS;
+        final long requestId = (clientRequestId == RequestTag.NON_EXISTENT_ID) ? idGenerator.get() : clientRequestId;
+
+        return new ContinuationTokenAsyncIterator<>(token -> {
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/server/SegmentHelper.java]: You have a lambda that contains an anonymous implementation of an interface. Let's refactor this code and make this a separate method, and the `FailingReplyProcessor` its own sub-class. That will make reading this much easier.

File: controller/src/main/java/io/pravega/controller/server/SegmentHelper.java [hunk 12]
```diff
@@ -1035,6 +1044,199 @@ public void authTokenCheckFailed(WireCommands.AuthTokenCheckFailed authTokenChec
...
+                                                                               final ConnectionFactory clientCF,
+                                                                               final String delegationToken,
+                                                                               final long clientRequestId) {
+
+        final Controller.NodeUri uri = getSegmentUri(scope, stream, 0L, hostControllerStore);
+        final String qualifiedName = getScopedStreamName(scope, stream);
+        final WireCommandType type = WireCommandType.READ_TABLE_ENTRIES;
+        final long requestId = (clientRequestId == RequestTag.NON_EXISTENT_ID) ? idGenerator.get() : clientRequestId;
+
+        return new ContinuationTokenAsyncIterator<>(token -> {
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/server/SegmentHelper.java]: Same here.

File: controller/src/main/java/io/pravega/controller/store/stream/Scope.java [hunk 13]
```diff
@@ -39,6 +43,35 @@
...
     CompletableFuture<Void> deleteScope();
 
+    /**
+     * Api to add stream under a scope for listing streams. 
+     * 
+     * @param name name of stream
+     * @param creationTime  creation time of stream.
+     * @return Future, which upon completion will indicate that the stream has been added to the scope. 
+     */
+    CompletableFuture<Void> addStreamToScope(final String name, final long creationTime);
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/store/stream/Scope.java]: How are these related to this pull request?

File: controller/src/main/java/io/pravega/controller/store/stream/Scope.java [hunk 14]
```diff
@@ -52,4 +85,26 @@
      */
     void refresh();
 
+    static String encodeStreamInScope(String name, long creationTime) {
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/store/stream/Scope.java]: These 3 methods would look much nicer if encapsulated (non-statically) into their own class:
- Two constructors: one with name and creation time and one with encoded name. Both constructors would immediately encode and store in a field.
- The class would implement comparable, which is what you have in `compareStreamInScope`.


Another question? Where is this encoding going? Is it stored anywhere? If so, have you considered versioning it?

File: controller/src/main/java/io/pravega/controller/store/stream/StreamMetadataStore.java [hunk 15]
```diff
@@ -189,6 +190,19 @@
...
 
+    /**
+     * List existing streams in scopes with pagination. This api continues listing streams from the supplied continuation token
+     * and returns a count limited list of streams and a new continuation token.
+     *
+     * @param scopeName Name of the scope
+     * @param continuationToken continuation token
+     * @param limit limit on number of streams to return.
+     * @param executor 
+     * @return A map of streams in scope to their configurations
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/store/stream/StreamMetadataStore.java]: This does not return a map.

File: controller/src/main/java/io/pravega/controller/store/stream/ZKScope.java [hunk 16]
```diff
@@ -40,13 +60,90 @@ public String getName() {
...
         return store.getChildren(scopePath);
     }
 
     @Override
-    public void refresh() {
+    public CompletableFuture<Pair<List<String>, String>> listStreamsInScope(int limit, String continuationToken, Executor executor) {
+        AtomicInteger set = new AtomicInteger(0);
+        AtomicReference<String> floor = new AtomicReference<>();
+        if (!Strings.isNullOrEmpty(continuationToken)) {
+            Pair<String, Long> decoded = Scope.decodeStreamInScope(continuationToken);
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/store/stream/ZKScope.java]: If this continuation token is external, it needs to be versioned, Please see my comment in `Scope.decodeStreamInScope`

File: controller/src/main/java/io/pravega/controller/store/stream/ZKScope.java [hunk 17]
```diff
@@ -40,13 +60,90 @@ public String getName() {
...
+    private int findSetForStream(String stream) {
+        int hash = stream.hashCode();
+
+        return hash == Integer.MIN_VALUE ? 0 : Math.abs(hash) % SET_COUNT;
     }
 
+    private CompletableFuture<List<String>> getStreamsInSet(Integer set, int limit, String continuationToken) {
+        String setPath = getSetPath(set);
+        return store.getChildren(setPath)
+                    .exceptionally(e -> {
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/store/stream/ZKScope.java]: Use `Futures.exceptionallyExpecting`

File: controller/src/main/java/io/pravega/controller/util/Config.java [hunk 18]
```diff
@@ -30,10 +30,10 @@
 @Slf4j
 public final class Config {
     private final static com.typesafe.config.Config CONFIG = ConfigFactory.defaultApplication()
-            .withFallback(ConfigFactory.defaultOverrides().resolve(ConfigResolveOptions.defaults().setAllowUnresolved(true)))
-            .withFallback(ConfigFactory.systemEnvironment())
-            .withFallback(ConfigFactory.defaultReference())
-            .resolve(ConfigResolveOptions.defaults().setAllowUnresolved(true));
+                                                                          .withFallback(ConfigFactory.defaultOverrides().resolve(ConfigResolveOptions.defaults().setAllowUnresolved(true)))
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/util/Config.java]: This change seems unnecessary

**Comment by andreipaduroiu** [comment]
I just realized this is built on top of #3312. I added the relevant comments there too.

File: controller/src/main/java/io/pravega/controller/store/stream/Scope.java (hunk 13 above)
Comment (shiveshr) [reply controller/src/main/java/io/pravega/controller/store/stream/Scope.java]: this is coming from issue #3278

File: controller/src/main/java/io/pravega/controller/store/stream/Scope.java (hunk 14 above)
Comment (shiveshr) [reply controller/src/main/java/io/pravega/controller/store/stream/Scope.java]: @shrids please update the PR with latest from PR #3312 as these methods have been removed.

File: controller/src/main/java/io/pravega/controller/store/stream/StreamMetadataStore.java (hunk 15 above)
Comment (shiveshr) [reply controller/src/main/java/io/pravega/controller/store/stream/StreamMetadataStore.java]: Fixed in the branch for #3312

File: controller/src/main/java/io/pravega/controller/store/stream/ZKScope.java (hunk 16 above)
Comment (shiveshr) [reply controller/src/main/java/io/pravega/controller/store/stream/ZKScope.java]: It has been made versioned in branch #3312

File: common/src/main/java/io/pravega/common/util/ContinuationTokenAsyncIterator.java (hunk 9 above)
Comment (shiveshr) [reply common/src/main/java/io/pravega/common/util/ContinuationTokenAsyncIterator.java]: answered on the original branch #3312 

@shrids lets have that branch merged before we share the review here as managing comments on two PRs for same change set becomes unwieldy.

File: controller/src/main/java/io/pravega/controller/server/SegmentHelper.java (hunk 11 above)
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/server/SegmentHelper.java]: On second thoughts, maybe we dont need to use ContinuationTokenAsync iterator here. 
Instead, you can have a simple iterator: 
```AsyncIterator<Map.Entry<IteratorState, Collection<TableSegment.IteratorItem<TableKey<byte[]>>>>>```
 that, as part of each getNext returns the pair of token and next batch of entries. 

That way you dont need to depend on other PR and the api here becomes explicit that it is giving out a token and a batch of entries rather than an "iterator".

Comment (shiveshr) [reply controller/src/main/java/io/pravega/controller/server/SegmentHelper.java]: the advantage is, if i want to get all keys, i can wrap it up in a continuation token Async Iterator. 
if i want to work only on a subset of keys, i can use the result as received by directly calling the api and working with a batch of results.. 
it will totally depend on the business logic that is interested in using these apis.. and there could be different ways it could be invoked in different places.

Comment (shrids) [reply controller/src/main/java/io/pravega/controller/server/SegmentHelper.java]: Thanks @shiveshr , will open a separate PR with the updated api.

**Comment by shrids** [comment]
Closing this PR. Created a new PR https://github.com/pravega/pravega/pull/3396 which does not use ContinuationTokenAsyncIterator.

**Comment by shrids** [comment]
[Issue 2895: Enable TableStore iterator apis on SegmentHelper](https://trello.com/c/95H650Mv/127-issue-2895-enable-tablestore-iterator-apis-on-segmenthelper)
